provider/simple/reprovide.go

```diff
@@ -0,0 +1,225 @@
+package simple
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	backoff "github.com/cenkalti/backoff"
+	cid "github.com/ipfs/go-cid"
+	cidutil "github.com/ipfs/go-cidutil"
+	blocks "github.com/ipfs/go-ipfs-blockstore"
+	pin "github.com/ipfs/go-ipfs/pin"
+	ipld "github.com/ipfs/go-ipld-format"
+	logging "github.com/ipfs/go-log"
+	merkledag "github.com/ipfs/go-merkledag"
+	verifcid "github.com/ipfs/go-verifcid"
+	routing "github.com/libp2p/go-libp2p-routing"
+)
+
+var logR = logging.Logger("reprovider.simple")
+
+//KeyChanFunc is function streaming CIDs to pass to content routing
+type KeyChanFunc func(context.Context) (<-chan cid.Cid, error)
+type doneFunc func(error)
+
+// Reprovider reannounces blocks to the network
+type Reprovider struct {
+	ctx     context.Context
+	trigger chan doneFunc
+
+	// The routing system to provide values through
+	rsys routing.ContentRouting
+
+	keyProvider KeyChanFunc
+
+	tick time.Duration
+}
+
+// NewReprovider creates new Reprovider instance.
+func NewReprovider(ctx context.Context, reprovideIniterval time.Duration, rsys routing.ContentRouting, keyProvider KeyChanFunc) *Reprovider {
+	return &Reprovider{
+		ctx:     ctx,
+		trigger: make(chan doneFunc),
+
+		rsys:        rsys,
+		keyProvider: keyProvider,
+		tick:        reprovideIniterval,
+	}
+}
+
+// Close the reprovider
+func (rp *Reprovider) Close() error {
+	return nil
+}
+
+// Run re-provides keys with 'tick' interval or when triggered
+func (rp *Reprovider) Run() {
+	// dont reprovide immediately.
+	// may have just started the daemon and shutting it down immediately.
+	// probability( up another minute | uptime ) increases with uptime.
+	after := time.After(time.Minute)
+	var done doneFunc
+	for {
+		if rp.tick == 0 {
+			after = make(chan time.Time)
+		}
+
+		select {
+		case <-rp.ctx.Done():
+			return
+		case done = <-rp.trigger:
+		case <-after:
+		}
+
+		//'mute' the trigger channel so when `ipfs bitswap reprovide` is called
+		//a 'reprovider is already running' error is returned
+		unmute := rp.muteTrigger()
+
+		err := rp.Reprovide()
+		if err != nil {
+			logR.Debug(err)
+		}
+
+		if done != nil {
+			done(err)
+		}
+
+		unmute()
+
+		after = time.After(rp.tick)
+	}
+}
+
+// Reprovide registers all keys given by rp.keyProvider to libp2p content routing
+func (rp *Reprovider) Reprovide() error {
```

Comment:
Should this be exported? That test should probably be calling `Trigger` instead.

Reply:
@Stebalien I just moved this code and the corresponding tests without changing them. It looks like, in order to call `Trigger` instead, the `Reprovider` needs to be `Run()` in the tests, which isn't currently happening. Are you OK just leaving this as it has been? I'm trying to get it deprecated anyway.

Reply:
SGTM.